File: src/ws/control.py
from .connection import Connection

class ControlConnection(Connection):
    def __init__(self, server, ws, user):
        super().__init__(ws, "client")
        self.server = server
        self.user = user

    async def handle(self, cmd, args):
        if cmd == "DEVICE":
            device_id, device_cmd, *device_args = args
            print(f"→ client {self.user.username} sent command to device {device_id}: {device_cmd} {' '.join(device_args)}")
            for conn in self.server.connections.values():
                if conn.type == "device" and conn.device.id == device_id:
                    return await conn.handle_from_client(self, device_cmd, device_args)
            return True
        return False

File: src/ws/control.ts
import { Connection } from "./connection.js";
import type { Server } from "./server.js";
import type { User } from "../data/models.js";

export class ControlConnection extends Connection {
  server: Server;
  user: User;

  constructor(server: Server, ws: any, user: User) {
    super(ws, "client");
    this.server = server;
    this.user = user;
  }

  async handle(cmd: string, args: string[]): Promise<boolean> {
    if (cmd !== "DEVICE") return false;

    const [deviceId, deviceCmd, ...deviceArgs] = args;
    if (!deviceId || !deviceCmd) return true;

    for (const conn of this.server.connections.values()) {
      if (conn.type === "device" && conn.device.id === deviceId) {
        return conn.handleFromClient(this, deviceCmd, deviceArgs);
      }
    }

    return true;
  }
}
